Message list

kusera opened the pull request:
##### TODO
- [x] Fetch existing messages from firebase
- [x] Change message list by clicking on a room
- [x] Send message to firebase
- [x] Message scroll to bottom on refresh and message received
- [x] Give focus to message input by hitting 'Enter' anywhere

Commit:
Add constants for message fetching

Commit:
Add some whitespace

Commit:
Add message fetching actions

Commit:
Add whitespace

Commit:
Move message converstion to MessageBox

Commit:
Add MessageStore and hook it up to dispatcher

Commit:
Hook up message retrieval from server

- MessageActions will be the bridge between MessageBox and ChatAPI,
  enabling the client to send messages to the server or fetch from the
  server
- The "currentRoom" is currently hardcoded into RoomStore
- Date dividers will eventually be dynamically inserted as messages are
  initially fetched from the server (will likely have a state of
  "lastMessageDate")

chikamichi commented:
It could be nice to add a short explanation about the feature you want to implement on that branch, in the PR's description.

chikamichi commented:
You may even use [tasks lists](https://help.github.com/articles/writing-on-github/#task-lists) to keep track of (and let people know about) your progress:

* [x] add the notion of a message list to the store
* [ ] display the messages

Commit:
Fix typos and clear up language

- Changed value of NEW_MESSAGE_FETCHED to 'NEW_MESSAGE_FETCHED' from
  'ONE_MESSAGE_FETCHED'
- Changed parameter of fetchRoomMessage from name to roomName

Commit:
Remove last cdn dependency

Commit:
Remove outdated and commented out code

Commit:
Remove '/addons' for consistency

Commit:
Change 'React' to 'react' everywhere

Commit:
Change feature list into a task list

- Removed the structure section since it wasn't being used

Commit:
Change indenting to break list formatting

Commit:
Fix formatting problems on github

Commit:
Revert changes because they look bad

Commit:
Add query to MessageStore for currentRoom

Commit:
Add '/addons' path back to requires

Commit:
Add message fetching by clicking rooms

MessageActions
- Moved fetching of messages out of MessageActions and into RoomActions
  because the action was reused to grab the room name

RoomActions
- Renamed arguments of RoomActions to be more descriptive

ServerActions
- Changed function names to reflect new constant names

ChatAPI
- Renamed ChatAPI functions to be more descriptive
- Repurposed getMessages to get room name as well

ChatController
- Removed a mysterious line of code with no use whatsoever

CreateRooModal
- Improved variable names
- Added an autofocus to the input

MessageBox
- Now holds the harded coded test room (in MessageStore before)

RoomBox and RoomNode
- Added injection for room changing

AppConstants
- Renamed constants to be more precise and descriptive

MessageStore
- Broke up the code into sections with commented headers
- Added functions which read and change the "current room"

RoomStore
- Broke up the code into sections with commented headers

Commit:
Marked off another feature

Commit:
Remove room name transformation for now

Commit:
Add message sending

MessageActions
- Add sendMessage action

ServerActions
- Repurposed roomMessagesFetched to messageFetched (fetches individual
  messages as children of room, rather than fetching the entire room)

ChatAPI
- Added function to send message to the firebase server
- Changed getMessages to listen for 'child_added' events
- Changed the message delivery mechanism to deliver them one by one
  (and thus recieve them one by one from the server)

MessageBox
- Added a function to convert text (received from MessageForm) into a
  JSON-like object to send to the server
- Add dependency injection to MessageForm
- Changed date conversion to account for time zones, server now uses
  UTC exclusively to date messages, but will be converted to local
  time once fetched

MessageForm
- Duplicated event handlers from CreateRoomModal to respond to user
  interaction
- 'Enter' to send not yet implemented

MessageInstance
- Moved style generation out of render() into its own function

AppConstants
- Now that messages are fetched as children rather than as a group,
  ROOM_MESSAGES_FETCHED is redundant was rolled along with
  NEW_MESSAGE_FETCHED into MESSAGE_FETCHED

MessageStore
- Changed to add messages one by one rather than as a block
- Appends the unique timestamp has generated by firebase an id
- Removed old outdated code for NEW_MESSAGE_FETCHED and rolled message
  receiving functions into a single case

Commit:
Add message delete listener for syncing

Commit:
Wipe current message list on room change

Commit:
Add focus message input via 'enter' globally

Commit:
Add auto scrolling to MessageList

Commit:
Lower the scrolling dead zone

Commit:
Strip all extra whitespace at end of messages

Commit:
Prevent 'refetching' from current room

Commit:
Prevent page reload on form submit via enter

Commit:
Improve functionality of createRoom and deleteRoom

- Create room and delete room now properly create the firebase objects
  necessary for storing messages and memberlist
- Also included a dummy message for new rooms
- Fixed broken variables from 4dcf887
- Improved argument names

Commit:
Fix gobal focus via enter
